src/xpcc/driver/position/vl53l0_impl.hpp

```diff
@@ -0,0 +1,633 @@
+{
+	if(measurementTime < TimeOverhead::Start + TimeOverhead::End) {
+		return false;
+	}
+
+	// available time for "final range" step in microseconds
+	uint32_t finalRangeTime = measurementTime - TimeOverhead::Start - TimeOverhead::End;
+	const uint16_t msrcDssTccTimeout = timeoutToMicroseconds(sequenceInfo.msrcDssTccTimeout,
+															 sequenceInfo.vcselPeriodPreRange);
+
+	if(sequenceInfo.enabledSteps & MeasurementSequenceStep::TCC) {
+		const uint16_t stepTime = TimeOverhead::TCC + msrcDssTccTimeout;
+		if(finalRangeTime <= stepTime) {
+			return false;
+		}
+		finalRangeTime -= stepTime;
+	}
+
+	// "dynamic spad selection" is enabled
+	if(sequenceInfo.enabledSteps & MeasurementSequenceStep::DSS) {
+		const uint16_t stepTime = 2 * (TimeOverhead::DSS + msrcDssTccTimeout);
+		if(finalRangeTime <= stepTime) {
+			return false;
+		}
+
+		finalRangeTime -= stepTime;
+	// additional time for msrc is only needed if dss is disabled
+	} else if(sequenceInfo.enabledSteps & MeasurementSequenceStep::MSRC) {
+		const uint16_t stepTime = TimeOverhead::MSRC + msrcDssTccTimeout;
+		if(finalRangeTime <= stepTime) {
+			return false;
+		}
+
+		finalRangeTime -= stepTime;
+	}
+
+	if(sequenceInfo.enabledSteps & MeasurementSequenceStep::PreRange) {
+		const uint16_t stepTime = TimeOverhead::PreRange
+			+ timeoutToMicroseconds(sequenceInfo.preRangeTimeout, sequenceInfo.vcselPeriodPreRange);
+
+		if(finalRangeTime <= stepTime) {
+			return false;
+		}
+
+		finalRangeTime -= stepTime;
+	}
+
+	if(finalRangeTime <= TimeOverhead::FinalRange) {
+		return false;
+	}
+
+	finalRangeTime -= TimeOverhead::FinalRange;
+
+	// convert value from microseconds to mclks
+	// from now on finalRangeTime is a value in mclks
+	finalRangeTime = timeoutToMclks(finalRangeTime, sequenceInfo.vcselPeriodFinalRange);
+
+	if(sequenceInfo.enabledSteps & MeasurementSequenceStep::PreRange) {
+		finalRangeTime += sequenceInfo.preRangeTimeout;
+	}
+
+	registerValue = encodeTimeout(finalRangeTime);
+	return true;
+}
+
+template < typename I2cMaster >
+uint32_t
+xpcc::Vl53l0<I2cMaster>::timeoutToMicroseconds(uint16_t mclks, uint8_t vcselPeriodPclks)
+{
+	// in nanoseconds
+	const uint32_t macroPeriodNs = ((UINT32_C(2304) * vcselPeriodPclks * 1655) + 500) / 1000;
+
+	return ((mclks * macroPeriodNs) + 500) / 1000;
+}
+
+template < typename I2cMaster >
+uint32_t
+xpcc::Vl53l0<I2cMaster>::timeoutToMclks(uint32_t microseconds, uint8_t vcselPeriodPclks)
+{
+	const uint32_t macroPeriodNs = ((UINT32_C(2304) * vcselPeriodPclks * 1655) + 500) / 1000;
+
+	return (((microseconds * 1000) + (macroPeriodNs / 2)) / macroPeriodNs);
+}
+
+template < typename I2cMaster >
+uint16_t
+xpcc::Vl53l0<I2cMaster>::encodeTimeout(uint32_t timeoutMclks)
+{
+	// Decompose timeoutMclks into the following expression:
+	// coefficient*(2^exponent) + 1
+
+	uint16_t encodedValue = 0;
+	uint32_t coefficient = 0;
+	uint16_t exponent = 0;
+
+	if (timeoutMclks > 0) {
```

Comment:
I probably started with `if(`, copied code from the Vl6180 driver which has `if (` and got finally confused by that. I'll unify the formatting.